File: src/main/java/com/vc/medium/L841.scala
package com.vc.medium

object L841 {
  def canVisitAllRooms(rooms: List[List[Int]]): Boolean = {
    val n = rooms.length
    val visited = new Array[Boolean](n)
    visited(0) = true
    def dfs(n: Int): Unit = {
      if(!visited(n)) {
        visited(n) = true
        rooms(n).foreach(key => dfs(key))
      }
    }
    rooms.head.foreach(key => dfs(key))
    !visited.contains(false)
  }
}

File: src/main/java/com/vc/medium/L841.java
package com.vc.medium;

import java.util.*;

class L841 {
    public boolean canVisitAllRooms(List<List<Integer>> rooms) {
        HashSet<Integer> visited = new HashSet<>();

        Queue<Integer> q = new LinkedList<>();
        q.offer(0);
        while(!q.isEmpty()) {
            int room = q.poll();

            if(visited.contains(room)) continue;

            visited.add(room);
            for(int keys: rooms.get(room)) q.offer(keys);
        }
        return visited.size() == rooms.size();
    }
}
